File: server/apps/nexotype/schemas/omics/protein_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class ProteinCreate(BaseModel):
    """Schema for creating a new protein"""
    transcript_id: int
    uniprot_accession: str
    sequence_aa: str


class ProteinUpdate(BaseModel):
    """Schema for updating a protein"""
    transcript_id: int = None
    uniprot_accession: str = None
    sequence_aa: str = None


# ==========================================
# Response Schemas
# ==========================================

class ProteinDetail(BaseModel):
    """Schema for protein details"""
    id: int
    transcript_id: int
    uniprot_accession: str
    sequence_aa: str
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class ProteinListResponse(BaseModel):
    """Response schema for listing proteins"""
    success: bool
    data: list[ProteinDetail] | None = None
    count: int = 0
    error: str | None = None


class ProteinResponse(BaseModel):
    """Response schema for single protein operations"""
    success: bool
    data: ProteinDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None

File: client/src/modules/nexotype/schemas/omics/protein.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const ProteinSchema = z.object({
  id: z.number(),
  transcript_id: z.number().int(),
  uniprot_accession: z.string().min(1).max(20),
  sequence_aa: z.string().min(1),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateProteinSchema = z.object({
  transcript_id: z.number().int(),
  uniprot_accession: z.string().min(1).max(20),
  sequence_aa: z.string().min(1),
});

export const UpdateProteinSchema = z.object({
  transcript_id: z.number().int().optional(),
  uniprot_accession: z.string().min(1).max(20).optional(),
  sequence_aa: z.string().min(1).optional(),
});

// ==========================================
// Types
// ==========================================

export type Protein = z.infer<typeof ProteinSchema>;
export type CreateProtein = z.infer<typeof CreateProteinSchema>;
export type UpdateProtein = z.infer<typeof UpdateProteinSchema>;

// ==========================================
// Response Types
// ==========================================

export type ProteinResponse = {
  success: boolean;
  data?: Protein;
  error?: string;
};

export type ProteinsResponse = {
  success: boolean;
  data?: Protein[];
  count?: number;
  error?: string;
};
